From: Michael Weghorn <m.weghorn@posteo.de>
Date: Sat, 23 May 2020 14:23:38 +0200
Subject: FileChooser: Add a 'current_filter' return value

Add a 'current_filter' element for the FileChooser
response which is used to specify which filter
was selected in the file chooser.

This will allow making Gtk's
'gtk_file_chooser_get_filter' work for the
portal native file chooser.

Bug: https://gitlab.gnome.org/GNOME/gtk/-/issues/1820
Origin: upstream, 1.7.3, commit:35fca7fae881bdaba1bebccf7775eba84407a488
---
 data/org.freedesktop.impl.portal.FileChooser.xml | 18 ++++++++++++++++++
 data/org.freedesktop.portal.FileChooser.xml      | 18 ++++++++++++++++++
 src/file-chooser.c                               |  5 +++++
 3 files changed, 41 insertions(+)

diff --git a/data/org.freedesktop.impl.portal.FileChooser.xml b/data/org.freedesktop.impl.portal.FileChooser.xml
index 873b2dd..8c684b0 100644
--- a/data/org.freedesktop.impl.portal.FileChooser.xml
+++ b/data/org.freedesktop.impl.portal.FileChooser.xml
@@ -109,6 +109,15 @@
               See org.freedesktop.portal.FileChooser.OpenFile() for details.
             </para></listitem>
           </varlistentry>
+          <varlistentry>
+            <term>current_filter (sa(us))</term>
+            <listitem>
+              <para>
+                The filter that was selected.
+                See org.freedesktop.portal.FileChooser.OpenFile() for details.
+              </para>
+            </listitem>
+          </varlistentry>
           <varlistentry>
             <term>writable b</term>
             <listitem><para>
@@ -216,6 +225,15 @@
               See org.freedesktop.portal.FileChooser.OpenFile() for details.
             </para></listitem>
           </varlistentry>
+          <varlistentry>
+            <term>current_filter (sa(us))</term>
+            <listitem>
+              <para>
+                The filter that was selected.
+                See org.freedesktop.portal.FileChooser.OpenFile() for details.
+              </para>
+            </listitem>
+          </varlistentry>
         </variablelist>
     -->
     <method name="SaveFile">
diff --git a/data/org.freedesktop.portal.FileChooser.xml b/data/org.freedesktop.portal.FileChooser.xml
index f0c7408..8582abc 100644
--- a/data/org.freedesktop.portal.FileChooser.xml
+++ b/data/org.freedesktop.portal.FileChooser.xml
@@ -151,6 +151,15 @@
             </para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term>current_filter (sa(us))</term>
+          <listitem>
+            <para>
+              The filter that was selected. This may match a filter in the
+              filter list or another filter that was applied unconditionally.
+            </para>
+          </listitem>
+        </varlistentry>
       </variablelist>
     -->
     <method name="OpenFile">
@@ -248,6 +257,15 @@
             See org.freedesktop.portal.FileChooser.OpenFile() for details.
           </para></listitem>
         </varlistentry>
+        <varlistentry>
+          <term>current_filter (sa(us))</term>
+          <listitem>
+            <para>
+              The filter that was selected.
+              See org.freedesktop.portal.FileChooser.OpenFile() for details.
+            </para>
+          </listitem>
+        </varlistentry>
       </variablelist>
     -->
     <method name="SaveFile">
diff --git a/src/file-chooser.c b/src/file-chooser.c
index 2e83821..166ae9a 100644
--- a/src/file-chooser.c
+++ b/src/file-chooser.c
@@ -79,6 +79,7 @@ send_response_in_thread_func (GTask        *task,
   const char **uris;
   GVariant *choices;
   gboolean for_save;
+  GVariant *current_filter;
 
   g_variant_builder_init (&results, G_VARIANT_TYPE_VARDICT);
   g_variant_builder_init (&ruris, G_VARIANT_TYPE_STRING_ARRAY);
@@ -100,6 +101,10 @@ send_response_in_thread_func (GTask        *task,
   if (choices)
     g_variant_builder_add (&results, "{sv}", "choices", choices);
 
+  current_filter = g_variant_lookup_value (options, "current_filter", G_VARIANT_TYPE ("(sa(us))"));
+  if (current_filter)
+    g_variant_builder_add (&results, "{sv}", "current_filter", current_filter);
+
   if (g_variant_lookup (options, "uris", "^a&s", &uris))
     {
       int i;
